collectors/sophossiem/collector.js

```diff
@@ -74,70 +77,152 @@ class SophossiemCollector extends PawsCollector {
-                        return callback(error.response.data);
+                    else if (error.response && error.response.status === 429) {
+                        state.poll_interval_sec = 900;
+                        AlLogger.info('API Request Limit Exceeded.');
+                        collector.reportApiThrottling(function () {
+                            return callback(null, [], state, state.poll_interval_sec);
+                        });
                     }
                     else {
-                        return callback(error);
+                        if (error && error.response && error.response.data) {
+                            error.response.data.errorCode = error.response.status;
+                            error.response.data.message = error.response.data.message ? error.response.data.message : error.response.message || '';
+                            return callback(error.response.data);
+                        }
+                        else {
+                            return callback(error);
+                        }
+                    }
+                });
+        }
+        else {
+            utils.authenticate(SOPHOS_AUTH_BASE_URL, clientId, clientSecret).then((token) => {
+                // while runing on live api server pass getTenantIdAndDataRegion hostname value "api.central.sophos.com"
+                utils.getTenantIdAndDataRegion(SOPHOS_API_BASE_URL, token).then((response) => {
+                    let tenantId = response.id;
+                    if (!tenantId) {
+                        return callback("TenantId not found.");
+                    }
+                    if (!response.apiHosts.dataRegion) {
+                        return callback("Please generate credentials for the tenant. Currently we do not support credentials for Organization and Partner.");
+                    }
+                    const apiHostsURL = response.apiHosts.dataRegion.replace(/^https:\/\/|\/$/g, '');
+                    let headers = {
+                        "X-Tenant-ID": tenantId,
+                        "Authorization": `Bearer ${token}`
+                    };
+                    utils.getAPILogs(apiHostsURL, headers, state, [], process.env.paws_max_pages_per_invocation)
+                        .then(({ accumulator, nextPage, has_more }) => {
+                            const newState = collector._getNextCollectionState(state, nextPage, has_more);
+                            AlLogger.info(`SIEM000002 Next collection in ${newState.poll_interval_sec} seconds`);
+                            return callback(null, accumulator, newState, newState.poll_interval_sec);
+                        }).catch((error) => {
+                            // set errorCode if not available in error object to showcase client error on DDMetric
+                            if (error.response && error.response.status === 401) {
```

Comment:
can we make common function for error handling and call in both token and api credential authentication